### PR TITLE
feat: Megamenu

### PR DESCRIPTION
**Pull request by mgax**
Fixes https://github.com/ietf-tools/wagtail_website/issues/326

- Rename class `MenuItem` to `SecondaryMenuItem`.
- Create a model to represent the new megamenu `MainMenuItem`. Populate with initial data based on the content of the former main menu.
- Render the megamenu using the Bootstrap [Dropdown component](https://getbootstrap.com/docs/5.3/components/dropdowns/), with a small amount of additional styling.
- Some light fixes to header markup and styling.
- Remove unused variables from template context and get the remaining ones to load lazily. This jumped out as a performance issue when logged in as admin – the wagtail userbar generates template context several times, so it's important for the menu content to load only when needed.

### Menu logic

- Each _Main menu item_ has an associated _Page_. It can be any page in the website.
  - If the page has Megamenu links (in its main section or secondary sections; see below), clicking it displays a Megamenu.
  - Otherwise, the menu item is a simple link to the page.
- _Main menu items_ that expand to Megamenus display the following content:
  - The page's _introduction_ field is shown on the left 1/3 column.
  - Links, grouped in sections, show on the right 2/3 column; each section has a heading, followed by links which flow in 2 equal columns.
    - Main section:
      - The heading is the title of the page associated with the _Main menu item_.
        - If the page is published, the heading is also a link to the page.
        - Otherwise, the heading is not a link.
      - Links to child pages (if they are published, and have the _Show in menus_ flag enabled, regardless of whether they are index pages or not).
    - Secondary sections: defined in the _Secondary sections_ streamfield of the _Main menu item_.
      - Each link may point to a _Page_ in the website or an external URL.
      - The link's label defaults to the page's _title_ but it may be overridden in the _title_ field.
- Initial data:
  - One _Main menu item_ for each page that was previously in the menu, in the same order.
  - For each child of that page, if any of its published sub-children have _Show in menus_ enabled, a section is created in _Secondary sections_, containing those sub-children.

### Screenshots
- An expanded megamenu
  Screenshot 2023-12-21 at 11 23 38
- The editor with preview
  Screenshot 2023-12-21 at 11 24 00

### CODE REVIEW

File: ietf/context_processors.py [hunk 1]
```diff
@@ -51,14 +39,11 @@ def social_menu(site):
 
 def global_pages(request):
     site = Site.find_for_request(request)
+    # XXX Return lazy values. This makes a big difference when a page renders
```
Comment (kesara) [inline ietf/context_processors.py:42]: What does XXX stand for?

File: ietf/context_processors.py [hunk 2]
```diff
@@ -51,14 +39,11 @@ def social_menu(site):
...
-        "SECONDARY_MENU": secondary_menu(site),
-        "SOCIAL_MENU": social_menu(site),
-        "BASE_URL": getattr(settings, "WAGTAILADMIN_BASE_URL", ""),
-        "DEBUG": getattr(settings, "DEBUG", ""),
-        "FB_APP_ID": getattr(settings, "FB_APP_ID", ""),
-        "SITE": site,
+        "HOME": lambda: home_page(site),
+        "MENU": lambda: get_main_menu(site),
+        "SECONDARY_MENU": lambda: secondary_menu(site),
+        "SOCIAL_MENU": lambda: social_menu(site),
```
Comment (kesara) [inline ietf/context_processors.py:48]: Doesn't this need to return `BLOG_INDEX` anymore?

**Comment by kesara** [comment]
@mgax In your PR description, did you forget to add the screenshots?

File: ietf/context_processors.py (hunk 1 above)
Comment (mgax) [reply ietf/context_processors.py:42]: It's a convention meaning "something unexpected is going on here and this is why".

File: ietf/context_processors.py (hunk 2 above)
Comment (mgax) [reply ietf/context_processors.py:48]: No, it was removed from templates in 53c3e66, and not referenced anywhere in the code now.

**Comment by mgax** [comment]
> @mgax In your PR description, did you forget to add the screenshots?

Yes indeed, fixed, thanks!

**Comment by kesara** [comment]
This change doesn't work with IAB website.
Because when IAB layout is selected sub menus don't work.
See:
(with IAB layout for IAB site)
Screenshot 2023-12-22 at 00 33 24

(with IETF layout for IAB site)
Screenshot 2023-12-22 at 00 33 34

New IAB site will have submenus as well: https://temporary.iab.org/

File: ietf/templates/includes/header_iab.html
Comment (kesara) [inline ietf/templates/includes/header_iab.html:1]: See https://github.com/ietf-tools/wagtail_website/pull/367#issuecomment-1866102474
The mega menu doesn't seem to work with the IAB site.

**Review by kesara — CHANGES_REQUESTED** [review]

**Comment by mgax** [comment]
> This change doesn't work with IAB website.
> Because when IAB layout is selected sub menus don't work.

That's weird, I've tested locally with a fresh restore of the IAB database, and the current _megamenu_ branch (commit 0b257a7), and the menu looks just like on https://temporary.iab.org.

On this PR, IAB used to be broken, but commit 398561d fixed it. Looking at the 2nd screenshot, _(with IETF layout for IAB site)_:

292192826-03395561-c418-4a0b-a5d0-905074e3f395

It seems there are `MainMenuItem` snippets in the database, which shouldn't be the case – commit 398561d changed the migration to only create `MainMenuItem` records if there is an `home.HomePage` page present, and the IAB database (at least, my copy of it) doesn't have one.

@kesara I suspect an earlier version of this branch was deployed to the IAB instance, so an older version of the migration `utils.0009_megamenu` was run. Could you please make sure the repository is at commit 0b257a7, then roll back the migration, and apply it again?

```shell
./manage.py migrate utils 0007
./manage.py migrate
```

**Comment by kesara** [comment]
> It seems there are `MainMenuItem` snippets in the database, which shouldn't be the case – commit [398561d](https://github.com/ietf-tools/wagtail_website/commit/398561d273b792b69d3c7b8f82dc2d5082bd7714) changed the migration to only create `MainMenuItem` records if there is an `home.HomePage` page present, and the IAB database (at least, my copy of it) doesn't have one.

@mgax I ran the up-to-date version, but I rand it as a multisite where I had IETF database and I created a separate *IAB Home* for testing. I'll go forward with the current changes. But it would be nice to have if mega menu that will work when there a multiple sites (both IETF & IAB varients) on the same instance.

File: ietf/templates/includes/header_iab.html
Comment (kesara) [reply ietf/templates/includes/header_iab.html:1]: See https://github.com/ietf-tools/wagtail_website/pull/367#issuecomment-1866779693

**Review by kesara — APPROVED** [review]

**Comment by mgax** [comment]
> But it would be nice to have a mega menu that will work when there a multiple sites (both IETF & IAB varients) on the same instance.

It _should_ work though. Perhaps the `if` that checks whether the current website has the IETF or IAB theme could be improved. The one [for the megamenu](https://github.com/ietf-tools/wagtail_website/blob/4f0031956aeab3e4b090277c6d7b3965a7b1b4da/ietf/utils/context_processors.py#L92-L96) is the same as the one which was already in place for [the home page and the secondary menu](https://github.com/ietf-tools/wagtail_website/blob/4f0031956aeab3e4b090277c6d7b3965a7b1b4da/ietf/context_processors.py#L10-L25). My suggestion would be to use `ietf.utils.models. LayoutSettings` to make the decision.

**Comment by kesara** [comment]
Yeah, `LayoutSettings` would be best things to be used.

**Pull request merged**